KEP-3857: Recursive Read-only (RRO) mounts

AkihiroSuda opened the pull request:
Depends on:
- [x] https://github.com/kubernetes/enhancements/pull/3858
- [x] https://github.com/kubernetes/kubernetes/pull/123272

Depended by:
- https://github.com/kubernetes/website/pull/45159

Relevant:
- https://github.com/kubernetes-sigs/cri-tools/pull/1344
- https://github.com/containerd/containerd/pull/9787

- - -
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
/kind api-change

#### What this PR does / why we need it:

Implement  KEP-3857: Recursive Read-only (RRO) mounts: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3857-rro-mounts
- https://github.com/kubernetes/enhancements/issues/3857
- https://github.com/kubernetes/enhancements/pull/3858

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
NONE

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Support Recursive Read-only (RRO) mounts  (KEP-3857)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/3857
```

- - -

An example manifest:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: rro
spec:
  volumes:
    - name: mnt
      hostPath:
        # tmpfs is mounted on /mnt/tmpfs
        path: /mnt
  containers:
    - name: busybox
      image: busybox
      args: ["sleep", "infinity"]
      volumeMounts:
        # /mnt-rro/tmpfs is not writable
        - name: mnt
          mountPath: /mnt-rro
          readOnly: true
          mountPropagation: None
          recursiveReadOnly: IfPossible
        # /mnt-ro/tmpfs is writable
        - name: mnt
          mountPath: /mnt-ro
          readOnly: true
        # /mnt-rw/tmpfs is writable
        - name: mnt
          mountPath: /mnt-rw
```

Requirements:
- Feature gate "RecursiveReadOnlyMounts" to be enabled
- Linux kernel >= 5.12
- runc >= 1.1

AkihiroSuda commented:
/test all

AkihiroSuda commented:
/label api-review

(Already reviewed in https://github.com/kubernetes/enhancements/pull/3858 though)

AkihiroSuda commented:
Added  e2e tests.

@kubernetes/sig-node-pr-reviews @kubernetes/sig-node-api-reviews @kubernetes/sig-node-leads PTAL

AkihiroSuda commented:
Split the CRI API part to a separate PR for ease of review and merge:
- https://github.com/kubernetes/kubernetes/pull/123272

AkihiroSuda commented:
/retest

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2156,6 +2156,31 @@ type VolumeMount struct {
 	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
+	// RecursiveReadOnly specifies recursive-readonly mode.
+	//
+	// 1. If ReadOnly is false, RecursiveReadOnly must be unspecified.
```
For API docs, we should turn this into plain english.  Something like:

```
RecursiveReadOnly specifies whether read-only mounts should be handled
recursively.

If ReadOnly is false, this field has no meaning and must be unspecified.

If ReadOnly is true, and this field is set to Disabled, the mount is not made
recursively read-only.  If this field is set to IfPossible, the mount is made
recursively read-only, if it is supported by the container runtime.  If this
field is set to Enabled, the mount is made recursively read-only if it is
supported by the container runtime, otherwise the pod will not be started and
an error will be generated to indicate the reason.

If this field is set to IfPossible or Enabled, MountPropagation must be set to
None (or be unspecified, which defaults to None).

If this field is not specified on a Pod it defaults to Disabled for backwards
compatibility.  If this field is not specified on an embedded PodSpec (e.g. in
a Deployment), it has no default value until a Pod is created, at which point
it will default to Disabled.
```

AkihiroSuda replied:
Applied the suggestion

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5656,6 +5719,24 @@ type NodeDaemonEndpoints struct {
 }
 
+// RuntimeClassFeatures is a set of runtime features.
+type RuntimeClassFeatures struct {
```
I get the need for this, but it almost feels like a different feature - do we need it for RRO?

AkihiroSuda replied:
The E2E test suite depends on this to detect whether the node supports RRO

haircommander replied:
I'm personally not a huge fan of adding it just for the tests. Can we detect whether the node supports by parsing the response to a create or something like that? Or wait until CRIs support the field and then add e2e tests?

AkihiroSuda replied:
Not just for tests. Useful for human users too. This was already discussed in the original KEP, but hard to find the link to the thread.

thockin reviewed (commented):
API wise this is mostly as expected, thanks.

AkihiroSuda commented:
/retest

AkihiroSuda commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5656,6 +5717,27 @@ type NodeDaemonEndpoints struct {
+	RecursiveReadOnlyMounts *bool `json:"recursiveReadOnlyMounts,omitempty" protobuf:"varint,1,opt,name=recursiveReadOnlyMounts"`
+	// UserNamespaces is set to true if the runtime class supports UserNamespaces.
+	// +optional
+	UserNamespaces *bool `json:"userNamespaces,omitempty" protobuf:"varint,2,opt,name=userNamespaces"`
```
(This field was not present in the original KEP, but added for consistency with the latest CRI API https://github.com/kubernetes/cri-api/commit/c3d03b1349992a948b39080898e4f9ed06429bbe)

haircommander replied:
is it possible to have an OR on the feature gates? Or maybe we just leave the RuntimeClassFeatures object ungated and gate the individual fields?

thockin replied:
We can but there has to be one release where the type is supported, but not used by default.  It's hard to reason about, so we probably shouldn't.

haircommander replied:
what would you suggest we do for this? we have an object whose two fields are only relevant to two different features

thockin replied:
Ideally we would have two different gates - one controlling the introduction of the RuntimeClass stuff, and the other on RRO.  Otherwise we just lump them together and you can't get RuntimeClassFeatures until RRO goes beta.  RRO seems like a low risk of delay and can maybe even jump straight to GA.

matthyx replied:
2 gates mean 2 KEPs for PRR - so keep that in mind :)

thockin replied:
not NECESSARILY but usually yes.

rphillips commented:
/triage accepted
/priority important-longterm

haircommander commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2328,3 +2361,68 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
+func runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName string, runtimeClasses []v1.RuntimeClass) bool {
+	for _, f := range runtimeClasses {
+		if f.Name == runtimeClassName {
+			if f.Features == nil {
+				return false
+			}
+			if f.Features.RecursiveReadOnlyMounts == nil {
+				return false
+			}
+			if !*f.Features.RecursiveReadOnlyMounts {
+				return false
+			}
+			return true
```
could this be
```
return f.Features != nil && f.Features.RecursiveReadOnlyMounts != nil && *f.Features.RecursiveReadOnlyMounts

```
I think I have the cases right but check my work:
Features == nil -> false
Features.RROM == nil -> false
*Features.RROM == false -> false

AkihiroSuda replied:
fixed

haircommander commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2328,3 +2361,68 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
+				m.Name, *m.MountPropagation)
+		}
+		return true, nil
+	case v1.RecursiveReadOnlyIfPossible:
+		propNone := true
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return featureGateEnabled && runtimeSupportsRRO && m.ReadOnly && propNone, nil
```
very picky nit: for some reason my brain wants this to be between disabled and enabled. I don't see it effecting the code flow. Feel free to ignore, it is of no consequence

AkihiroSuda replied:
fixed

mrunalp commented:
Aren't we missing the logic to drop/disable the field in API if the feature gate is not enabled? cc: @thockin

AkihiroSuda commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -217,18 +217,21 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus, handlers []*runtimeap
-			Name:                   h.Name,
-			SupportsUserNamespaces: supportsUserns,
+		// The runtime features are exposed regardless to the values of the Kubernetes feature gates.
+		// https://github.com/kubernetes/kubernetes/pull/123216#discussion_r1484762897
```
cc @mrunalp

mrunalp replied:
I am testing with this PR on a local cluster without the feature gate enabled and I see that:
1. If I use this example
```
apiVersion: v1
kind: Pod
metadata:
  name: rro
spec:
  volumes:
    - name: mnt
      hostPath:
        # tmpfs is mounted on /mnt/tmpfs
        path: /mnt
  containers:
    - name: busybox
      image: busybox
      args: ["sleep", "infinity"]
      volumeMounts:
        # /mnt-rro/tmpfs is not writable
        - name: mnt
          mountPath: /mnt-rro
          readOnly: true
          mountPropagation: None
          recursiveReadOnly: IfPossible
```

then the pod goes to running with the status of rro as Disabled. (See my other comment on the code).

2. If the Feature Gate isn't enabled for User Namespaces for example, then we do something like this:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L552C1-L559C3

```
	// If the feature is disabled and not in use, drop the hostUsers field.
	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) && !hostUsersInUse(oldPodSpec) {
		// Drop the field in podSpec only if SecurityContext is not nil.
		// If it is nil, there is no need to set hostUsers=nil (it will be nil too).
		if podSpec.SecurityContext != nil {
			podSpec.SecurityContext.HostUsers = nil
		}
	}
```

mrunalp replied:
I will defer to API review (@thockin ) for this.

AkihiroSuda replied:
Updated `dropDisabledFields`

mrunalp commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2345,3 +2361,59 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return featureGateEnabled && runtimeSupportsRRO && m.ReadOnly && propNone, nil
```
Should we error here if:
1. Feature gate is not set OR
2. ReadOnly is not set OR
3. Mount Prop is not none
? i.e. IfPossible should only be tied to runtime support.

AkihiroSuda replied:
I expect "IfPossible" to make the best effort to avoid errors, and just fall back to "Disabled"

AkihiroSuda replied:
Because the IfPossible mode was introduced to allow creating a manifest that keeps compatibility on every node.
It ruins the purpose of IfPossible if we make it fallable.

haircommander commented on test/e2e_node/mount_rro_linux_test.go:
```diff
@@ -0,0 +1,434 @@
+	return false
+}
+
+var _ = SIGDescribe("Mount recursive read-only [LinuxOnly]", framework.WithSerial(), nodefeature.RecursiveReadOnlyMounts, func() {
```
nit: i usually prefer calling functions to be above called functions so the file can be read top-down. If you need to address more notes, would you mind reordering?

AkihiroSuda replied:
fixed

haircommander commented:
kubelet pieces
/lgtm

cc @thockin @mrunalp

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

thockin commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -343,13 +343,19 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 		klog.V(5).InfoS("Mount has propagation", "pod", klog.KObj(pod), "containerName", container.Name, "volumeMountName", mount.Name, "propagation", propagation)
 		mustMountRO := vol.Mounter.GetAttributes().ReadOnly
 
+		rro, err := resolveRecursiveReadOnly(mount, supportsRRO, utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts))
```
This is not the normal pattern for gates - is there a reason?  Usually we would check the gate and wipe the field in  pkg/api/pod/util.go `dropDisabledFields()`.

thockin replied:
Also, the gate needs to control ALL of the new fields, including status.

AkihiroSuda replied:
`resolveRecursiveReadOnly` takes a gate as an argument for ease of unit testing

thockin replied:
First and foremost, the gate prevents the API from being used by net-new objects.  When a user specifies RRO and the gate is disabled, we erase the field and pretend we don't know about it.  If RRO was already in use, we allow it to stay in use.

It's a secondary question whether we actually respect RRO.  Superficially, it seems reasonable to say "I know it says RRO, but that was from before, and we're not going to respect that".

AkihiroSuda replied:
Updated `dropDisabledFields`

haircommander commented:
> Is it OK that this is behind the same gate?

I think it should be behind +featureGate=UserNamespacesSupport

AkihiroSuda commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -554,3 +554,23 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
 	return autoConvert_v1_PersistentVolumeSpec_To_core_PersistentVolumeSpec(in, out, s)
 }
+
+// Convert_core_ContainerStatus_To_v1_ContainerStatus is defined outside the autogenerated file for use by other API packages
```
`make update` didn't generate this, and the files couldn't be compiled without manually adding this

thockin replied:
That's odd - I just checked out your PR and removed those and ran `hack/update-codegen.sh` and it works

AkihiroSuda replied:
Removed.

I guess something was temporarily broken when I tried a couple of weeks ago.

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5672,6 +5733,27 @@ type NodeDaemonEndpoints struct {
+}
+
+// RuntimeClass is a set of runtime class information.
+type RuntimeClass struct {
```
nit: this should be NodeRuntimeClass (as the type name).  Similar for other types.

AkihiroSuda replied:
done

AkihiroSuda commented on pkg/apis/core/types.go:
```diff
@@ -2698,6 +2735,10 @@ type ContainerStatus struct {
 	Resources *ResourceRequirements
+	// Status of volume mounts.
+	// +listType=atomic
+	// +optional
```
done

AkihiroSuda commented on pkg/apis/core/types.go:
```diff
@@ -4845,6 +4924,9 @@ type NodeStatus struct {
 	// Status of the config assigned to the node via the dynamic Kubelet config feature.
 	// +optional
 	Config *NodeConfigStatus
+	// The available runtime classes.
```
done

AkihiroSuda commented on pkg/api/pod/util.go:
```diff
@@ -776,6 +794,18 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
+		for i := range podStatus.EphemeralContainerStatuses {
+			podStatus.EphemeralContainerStatuses[i].VolumeMounts = nil
+		}
+	}
```
Drop doesn't seem working for podStatus, although it is working for podSpec

AkihiroSuda commented on pkg/registry/core/node/strategy.go:
```diff
@@ -101,6 +103,9 @@ func dropDisabledFields(node *api.Node, oldNode *api.Node) {
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
+		node.Status.RuntimeClasses = nil
+	}
```
This doesn't seem working

AkihiroSuda commented on pkg/api/pod/util.go:
```diff
@@ -671,6 +671,24 @@ func dropDisabledFields(
+				podSpec.EphemeralContainers[i].VolumeMounts[j].RecursiveReadOnly = nil
+			}
+		}
+	}
```
(Now, `recursiveReadOnly: Enabled` is just silently ignored when the feature gate is disabled)

AkihiroSuda commented on test/e2e_node/mount_rro_linux_test.go:
```diff
@@ -0,0 +1,440 @@
+				if initialConfig.FeatureGates == nil {
+					initialConfig.FeatureGates = make(map[string]bool)
+				}
+				initialConfig.FeatureGates[string(kubefeatures.RecursiveReadOnlyMounts)] = true
```
(How to set this for apiserver too?)

AkihiroSuda replied:
This seems to work
```
make test-e2e-node TEST_ARGS='--service-feature-gates=RecursiveReadOnlyMounts=true --kubelet-flags="--feature-gates=RecursiveReadOnlyMounts=true"' FOCUS="Mount recursive read-only" SKIP=""
```

MaryamTavakkoli commented:
/milestone v1.30

mrunalp commented:
/milestone v1.30

AkihiroSuda commented:
/retest

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2156,6 +2156,29 @@ type VolumeMount struct {
+	// If this field is set to IfPossible or Enabled, MountPropagation must be set to
+	// None (or be unspecified, which defaults to None).
+	//
+	// If this field is not specified on a Pod it defaults to Disabled for backwards
```
This is tricky.  We *want* to add static defaulting logic, but we have other issues around Kubelet that have given us a moratorium on new fields with default values.  So we *can't* add a static default here, which means the implementation code has to handle `nil` and `"Disabled"` as equivalent.

AkihiroSuda replied:
Reworded to ` If this field is not specified, it is treated as an equivalent of Disabled.`

AkihiroSuda commented on pkg/api/pod/util.go:
```diff
@@ -1098,6 +1128,28 @@ func clusterTrustBundleProjectionInUse(podSpec *api.PodSpec) bool {
+	if podSpec == nil {
+		return false
+	}
+	for _, t := range []ContainerType{InitContainers, Containers, EphemeralContainers} {
```
fixed

AkihiroSuda commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2156,6 +2156,29 @@ type VolumeMount struct {
+	// RecursiveReadOnly specifies whether read-only mounts should be handled
+	// recursively.
+	//
+	// If ReadOnly is false, this field has no meaning and must be unspecified.
```
How should the `PodValidationOptions` field look like?

AkihiroSuda replied:
Added commit `api: validate RecursiveReadOnlyMounts`

AkihiroSuda commented:
/retest

AkihiroSuda commented:
The deadline is approaching in 12 hours (18:00 PST Friday, 8th March 2024)
https://groups.google.com/g/kubernetes-sig-release/c/XrXHzMlyHWw

haircommander commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2131,6 +2144,16 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 				}
 				if vol.ReadOnly {
 					rroMode := v1.RecursiveReadOnlyDisabled
+					if b, err := resolveRecursiveReadOnly(vol, supportsRRO); err != nil {
```
can `b` be `isRRO` or something? it can still be short, but b confused me at first

AkihiroSuda replied:
It is normal to have a single-character variable in Go.
I also feel `isRRO` is more confusing when there are `rroMode` and `supportsRRO`. Just my feeling though.

AkihiroSuda replied:
`supportsRRO` is already in use, so let me keep it `b`.
Refactoring can be happen in a follow-up PR if you still like.

mrunalp commented on pkg/apis/core/types.go:
```diff
@@ -2011,6 +2011,26 @@ type VolumeMount struct {
+	//
+	// If this field is not specified, it is treated as an equivalent of Disabled.
+	//
+	// +featureGate=RecursiveReadOnlyMounts
```
We are referring to the featureGate before it is introduced in the next commit.

AkihiroSuda replied:
fixed

mrunalp commented on pkg/features/kube_features.go:
```diff
@@ -968,6 +968,13 @@ const (
+	// alpha: v1.30
+	//
+	// Allows recursive read-only mounts.
+	RecursiveReadOnlyMounts featuregate.Feature = "RecursiveReadOnlyMounts"
```
I think this whole commit should be moved up before we use it.

AkihiroSuda replied:
fixed

mrunalp commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2420,3 +2443,55 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return runtimeSupportsRRO && m.ReadOnly && propNone, nil
```
I still think that we should return errors in case of readonly true or mountProp not set to None because those fields are in control of the user writing the pod spec. If their runtime actually supports RRO they would miss it when we have an opportunity to return an error so they can correct it and get RRO.

AkihiroSuda replied:
fixed

mrunalp commented:
I think I have only one major outstanding comment otherwise the node bits look fine to me.

AkihiroSuda committed:
kubelet: change map[string]RuntimeHandler to []RuntimeHandler

The map is changed to an array so as to retain the order of the original array
propagated from the CRI runtime.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: RuntimeHandler: add SupportsRecursiveReadOnlyMounts

For KEP-3857: Recursive Read-only (RRO) mounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
pkg/features: add RecursiveReadOnlyMounts

For KEP-3857: Recursive Read-only (RRO) mounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
pod: dropDisabledFields: recognize RecursiveReadOnlyMounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
api: KEP-3857: Recursive Read-only (RRO) mounts

This commit modifies the following files:

- pkg/apis/core/types.go
- staging/src/k8s.io/api/core/v1/types.go

Other changes were auto-generated by running `make update`.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
node: dropDisabledFields: recognize RecursiveReadOnlyMounts gate

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
api: validate RecursiveReadOnlyMounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: expose `node.status.runtimeClasses`

For KEP-3857: Recursive Read-only (RRO) mounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: expose `containerStatuses.volumeMounts`

For KEP-3857: Recursive Read-only (RRO) mounts

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: KEP-3857: Recursive Read-only (RRO) mounts

See <https://kep.k8s.io/3857>.

An example manifest:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: rro
spec:
  volumes:
    - name: mnt
      hostPath:
        # tmpfs is mounted on /mnt/tmpfs
        path: /mnt
  containers:
    - name: busybox
      image: busybox
      args: ["sleep", "infinity"]
      volumeMounts:
        # /mnt-rro/tmpfs is not writable
        - name: mnt
          mountPath: /mnt-rro
          readOnly: true
          mountPropagation: None
          recursiveReadOnly: IfPossible
        # /mnt-ro/tmpfs is writable
        - name: mnt
          mountPath: /mnt-ro
          readOnly: true
        # /mnt-rw/tmpfs is writable
        - name: mnt
          mountPath: /mnt-rw
```

Requirements:
- Feature gate "RecursiveReadOnlyMounts" to be enabled
- Linux kernel >= 5.12
- runc >= 1.1

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
e2e: KEP-3857: Recursive Read-only (RRO) mounts

Usage:
```
make test-e2e-node \
  TEST_ARGS='--service-feature-gates=RecursiveReadOnlyMounts=true --kubelet-flags="--feature-gates=RecursiveReadOnlyMounts=true"' \
  FOCUS="Mount recursive read-only" SKIP=""
```

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
> I think I have only one major outstanding comment otherwise the node bits look fine to me.

Fixed, PTAL 🙏

mrunalp commented:
@thockin for API approval / lgtm if around

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1321,6 +1321,32 @@ func validateMountPropagation(mountPropagation *core.MountPropagationMode, conta
+		return nil
+	}
+	allErrs := field.ErrorList{}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
```
You don't need to check the gate here - if the dropDisabled logic left the field in place, we validate it.

AkihiroSuda replied:
fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1321,6 +1321,32 @@ func validateMountPropagation(mountPropagation *core.MountPropagationMode, conta
+		// NOP
+	case core.RecursiveReadOnlyEnabled, core.RecursiveReadOnlyIfPossible:
+		if !mount.ReadOnly {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "recursive read-only mount mode is available only to read-only mounts"))
```
nit: just "may only be specified when readOnly is true"

AkihiroSuda replied:
fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1321,6 +1321,32 @@ func validateMountPropagation(mountPropagation *core.MountPropagationMode, conta
+			allErrs = append(allErrs, field.Forbidden(fldPath, "recursive read-only mount mode is available only to read-only mounts"))
+		}
+		if mount.MountPropagation != nil && *mount.MountPropagation != core.MountPropagationNone {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "recursive read-only mount mode require propagation to be None or unset"))
```
"may only be specified when mountPropagation is None or not specified"

AkihiroSuda replied:
fixed

thockin commented on pkg/api/pod/util.go:
```diff
@@ -385,6 +400,7 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowInvalidTopologySpreadConstraintLabelSelector: false,
 		AllowNamespacedSysctlsForHostNetAndHostIPC:        false,
 		AllowNonLocalProjectedTokenPath:                   false,
+		AllowInvalidRecursiveReadOnlyMounts:               false,
```
I think you don't need this at all - dropDisabled already wipes it

AkihiroSuda replied:
fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2909,6 +2935,9 @@ func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]strin
 		if mnt.MountPropagation != nil {
 			allErrs = append(allErrs, validateMountPropagation(mnt.MountPropagation, container, fldPath.Child("mountPropagation"))...)
 		}
+		if !opts.AllowInvalidRecursiveReadOnlyMounts && mnt.RecursiveReadOnly != nil {
```
Don't need the conditional `AllowInvalidRecursiveReadOnlyMounts` - if the field has a value at this point, we validate it.

AkihiroSuda replied:
fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1321,6 +1321,32 @@ func validateMountPropagation(mountPropagation *core.MountPropagationMode, conta
+			allErrs = append(allErrs, field.Forbidden(fldPath, "recursive read-only mount mode require propagation to be None or unset"))
+		}
+	default:
+		allErrs = append(allErrs, field.InternalError(fldPath,
```
you want `field.NotSupported`

AkihiroSuda replied:
fixed

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2866,7 +2892,7 @@ func GetVolumeDeviceMap(devices []core.VolumeDevice) map[string]string {
 }
 
-func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]string, volumes map[string]core.VolumeSource, container *core.Container, fldPath *field.Path) field.ErrorList {
+func ValidateVolumeMounts(mounts []core.VolumeMount, voldevices map[string]string, volumes map[string]core.VolumeSource, container *core.Container, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
if you drop the conditional, then you don't need this extra arg, which removes deltas from this PR

AkihiroSuda replied:
fixed

thockin reviewed (commented):
I can live with these as a followup, but @kubernetes/release-team has final say

thockin commented:
Thanks!

/lgtm
/approve

AkihiroSuda commented:
Thanks for reviewing and approval, could you review the docs PR too?
- https://github.com/kubernetes/website/pull/45159
